libbeat/autodiscover/providers/kubernetes/service.go

```diff
@@ -33,44 +35,59 @@ import (
 )
 
 type service struct {
-	uuid    uuid.UUID
-	config  *Config
-	metagen kubernetes.MetaGenerator
-	logger  *logp.Logger
-	publish func(bus.Event)
-	watcher kubernetes.Watcher
+	uuid             uuid.UUID
+	config           *Config
+	metagen          metadata.MetaGen
+	logger           *logp.Logger
+	publish          func(bus.Event)
+	watcher          kubernetes.Watcher
+	namespaceWatcher kubernetes.Watcher
 }
 
 // NewServiceEventer creates an eventer that can discover and process service objects
 func NewServiceEventer(uuid uuid.UUID, cfg *common.Config, client k8s.Interface, publish func(event bus.Event)) (Eventer, error) {
-	metagen, err := kubernetes.NewMetaGenerator(cfg)
-	if err != nil {
-		return nil, err
-	}
-
 	logger := logp.NewLogger("autodiscover.service")
 
 	config := defaultConfig()
-	err = cfg.Unpack(&config)
+	err := cfg.Unpack(&config)
 	if err != nil {
 		return nil, err
 	}
 
 	watcher, err := kubernetes.NewWatcher(client, &kubernetes.Service{}, kubernetes.WatchOptions{
 		SyncTimeout: config.SyncPeriod,
-	})
+		Namespace:   config.Namespace,
+	}, nil)
 
 	if err != nil {
 		return nil, fmt.Errorf("couldn't create watcher for %T due to error %+v", &kubernetes.Service{}, err)
 	}
 
+	var namespaceMeta metadata.MetaGen
+	var namespaceWatcher kubernetes.Watcher
```

Comment:
do we really need an extra watcher to retrieve namespace details? I would hope this can be done through cached calls